src/server/game/Movement/PathGenerator.cpp

```diff
@@ -915,35 +915,121 @@ void PathGenerator::ReducePathLenghtByDist(float dist)
+        This means the distance from the target to the path's last point AFTER CUT can be higher than dist.
+        3) The path's last point AFTER CUT can be out of LoS from the target.
+
+        We need to make sure that in ANY case: 
+        a) the last point after cut is at MAX at "dist" yards from the target + marginal error factor (0.5f)
+        b) the last point after cut is in LoS of the target
+
+        In order to do that, we iterate on the points of the path starting by the point on idx+1 and we cut the path at the first
+        point that is both in LoS and in range.
+        Worst case scenario, we use the last position, which should be close (if not equal) to the target position.
+        */
+        if (IsValidFinalPoint(newFinalPoint, target, dist))
+        {
+            _pathPoints[idx + 1] = newFinalPoint;
+            _pathPoints.resize(idx + 2);
         }
-        else if (i == 0) // at second point
+        else
+        {
+            CutAtFirstValidPoint(idx + 1, target, dist);
+        }
+    }
+    else if (totalLengthSoFar == dist && idx > 0)
+    {
+        if (IsValidFinalPoint(_pathPoints[idx], target, dist))
+        {
+            _pathPoints.resize(idx + 1);
+        }
+        else
+        {
+            CutAtFirstValidPoint(idx + 1, target, dist);
+        }
+    }
+    else
+    {
+        // if initial position is "valid", no need to move
+        if (IsValidFinalPoint(_pathPoints[0], target, dist))
         {
             _pathPoints[1] = _pathPoints[0];
             _pathPoints.resize(2);
-            break;
         }
-
-        dist -= len;
-        nextVec = currVec; // we're going backwards
+        else
+        {
+            CutAtFirstValidPoint(1, target, dist);
+        }
     }
 }
 
 bool PathGenerator::IsInvalidDestinationZ(Unit const* target) const
 {
     return (target->GetPositionZ() - GetActualEndPosition().z) > 5.0f;
 }
+
+bool PathGenerator::IsValidFinalPoint(const G3D::Vector3 finalPoint, Unit* target, float dist) const
+{
+    return target->GetExactDist(finalPoint.x, finalPoint.y, finalPoint.z) <= dist + 0.5f
+        && target->IsWithinLOS(finalPoint.x, finalPoint.y, finalPoint.z);
+}
+
+uint32 PathGenerator::SearchForFirstValidPoint(int startAtIndex, Unit* target, float dist) const
+{
+    uint32 index = startAtIndex;
+    for ( ; index < _pathPoints.size() ; index++)
+    {
+        if (IsValidFinalPoint(_pathPoints[index], target, dist))
+            break;
+    }
+    return index;
+}
+
+void PathGenerator::CutAtFirstValidPoint(int startAtIndex, Unit* target, float dist)
+{
+    // directly test the last point in order to avoid iterating over all points in case of unreachable target
+    if (!IsValidFinalPoint(_pathPoints[_pathPoints.size() - 1], target, dist))
+    {
+        // even the last position of the generated path, which should be close if not on top of the target, was not good enough. we mark the path as unusable.
+        _type = PATHFIND_NOPATH;
```

Comment:
this code doesn't handle PATHFIND_SHORT case (and it's still to be defined after which distance it should be set to PATHFIND_SHORT)